src/honey_badger.rs

```diff
@@ -577,37 +563,96 @@ pub enum MessageContent<NodeUid> {
     },
 }
 
-impl<NodeUid> MessageContent<NodeUid> {
-    fn with_epoch(self, epoch: u64) -> Message<NodeUid> {
-        Message {
+/// A message sent to or received from another node's Honey Badger instance.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Message<NodeUid> {
+    epoch: u64,
+    content: MessageContent<NodeUid>,
+}
+
+impl<NodeUid> Message<NodeUid> {
+    pub fn epoch(&self) -> u64 {
+        self.epoch
+    }
+}
+
+/// Content of a message with internal optimisations to reduce cloning inside `HoneyBadger`.
+pub enum InternalMessageContent<NodeUid>
+where
+    NodeUid: Clone + Debug + for<'de> Deserialize<'de> + Serialize,
+{
+    /// A message belonging to the common subset algorithm in the given epoch.
+    CommonSubset(common_subset::Message<NodeUid>),
+    /// A decrypted share of the output of `proposer_id`.
+    DecryptionShare {
+        proposer_id: NodeUid,
+        share: Rc<DecryptionShare<Bls12>>,
+    },
+}
+
+impl<NodeUid> InternalMessageContent<NodeUid>
+where
+    NodeUid: Clone + Debug + for<'de> Deserialize<'de> + Serialize,
+{
+    fn with_epoch(self, epoch: u64) -> InternalMessage<NodeUid> {
+        InternalMessage {
             epoch,
             content: self,
         }
     }
 }
 
-/// A message sent to or received from another node's Honey Badger instance.
-#[derive(Clone, Debug, Deserialize, Serialize)]
-pub struct Message<NodeUid> {
+/// The type of internal message. An internal message has to be externalised, that is, converted to
+/// a `Message` before being sent out.
+pub struct InternalMessage<NodeUid>
+where
+    NodeUid: Clone + Debug + for<'de> Deserialize<'de> + Serialize,
+{
     epoch: u64,
-    content: MessageContent<NodeUid>,
+    content: InternalMessageContent<NodeUid>,
 }
 
-impl<NodeUid> Message<NodeUid> {
+impl<NodeUid> InternalMessage<NodeUid>
+where
+    NodeUid: Clone + Debug + for<'de> Deserialize<'de> + Serialize,
+{
     pub fn epoch(&self) -> u64 {
         self.epoch
     }
+
+    pub fn into_external(self) -> Message<NodeUid> {
+        let InternalMessage { epoch, content } = self;
+        Message {
+            epoch,
+            content: match content {
+                InternalMessageContent::CommonSubset(msg) => MessageContent::CommonSubset(msg),
+                InternalMessageContent::DecryptionShare { proposer_id, share } => {
+                    // FIXME: Handle errors.
+                    let share_copy = Rc::try_unwrap(share).unwrap();
```

Comment:
I think that _will_ usually fail: Most of the time, the local reference will live longer than the message.

I hadn't realized that `Rc` doesn't automatically implement the serde traits. Maybe instead of the internal-external distinction the message should just always have an `Rc`, but we should implement `Serialize` and `Deserialize` manually? (Or rather with some attribute `#[serde(with = "a_new_rc_serde_module")]`.)

But maybe this is all premature optimization; I'm happy to postpone that to a separate PR.

Reply:
Issue #77